zrml/neo-swaps/src/lib.rs

```diff
@@ -979,4 +979,89 @@ mod pallet {
             Self::do_deploy_pool(who, market_id, amount, spot_prices, swap_fee)
         }
     }
+
+    impl<T: Config> Pallet<T> {
+        fn amount_including_fee_surplus(
+            amount: BalanceOf<T>,
+            fee_fractional: BalanceOf<T>,
+        ) -> Result<BalanceOf<T>, DispatchError> {
+            let fee_divsisor = ZeitgeistBase::<BalanceOf<T>>::get()?
+                .checked_sub(&fee_fractional)
+                .ok_or(Error::<T>::Unexpected)?;
+            amount.bdiv(fee_divsisor)
+        }
+
+        fn total_fee_fractional(
+            swap_fee: BalanceOf<T>,
+            external_fee_percentage: Perbill,
+        ) -> Result<BalanceOf<T>, DispatchError> {
+            let external_fee_fractional =
+                external_fee_percentage.mul_ceil(ZeitgeistBase::<BalanceOf<T>>::get()?);
```

Comment:
Can using `mul_ceil` here result in overcharging? Thinking about the scenario where a user wants to place a bet using his whole balance and the frontend calculates the swap amounts, but is off by one because of this and the TX fails.

Reply:
`mul_ceil` rounds up, which results in the formula `buy_amount = amount / (1 - fee)` to have a higher buy amount than calculated on the AMM. Yes, we should round down here. Good catch!

Reply:
https://github.com/zeitgeistpm/zeitgeist/pull/1275/commits/9731b37fd3d4101cc6176910abffc1ffeaf7b033